### PR TITLE
206feature/file storage provider

### PR DESCRIPTION
**Pull request by esteban-gs**
## Dev Summary

- File storage service, implements either azure storage for production mode or local disk in development. The service resolves the provider based on the `MODE` env var
- New profile controller with an auth guard to extract the current user from the cookie. Saves the file and stores the url in the user table
- New entity property for the profile url

## Test Plan

repro steps:

1. sign in with postman, and it should store the cookie
2. make a PUT request with multi-form body:
```bash
curl --request PUT \
  --url http://localhost:3001/api/auth/users/profile/12 \
  --header 'Content-Type: multipart/form-data; boundary=---011000010111000001101001' \
  --cookie 'NEH_is_cool=s%253AeyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTIsImZpcnN0TmFtZSI6InVzZXIxRmlyc3QiLCJsYXN0X25hbWUiOiJ1c2VyMUxhc3QiLCJlbWFpbCI6InVzZXIxRmlyc3RAZXhhbXBsZS5jb20iLCJlbWFpbF9ub3RpZmljYXRpb25fb3B0X291dCI6ZmFsc2UsInByb2ZpbGVfaW1hZ2VfdXJsIjoiaHR0cDovL2xvY2FsaG9zdDozMDAwL3VzZXJwcm9maWxlLTEyLnBuZyIsImlhdCI6MTY3NTk5MjYzMSwiZXhwIjoxNjc1OTk2MjMxfQ.8qVpZq1cZRyLVW2sqEtOTFefGiWJMko1MzTP00SGNOY.KDGi8EYHaQsImli12p2q616NC2%252BgHrwQH2pp3%252FZKO%252BU; .AspNetCore.Antiforgery.4gFydmEKsEU=CfDJ8DlqRFsOpwhFsCNlaNRzxq7LhgKekJCsdzM0QeDsHFUr5l6cMmFPsWnskpl62WtPd_RZ3TxQixaPOUju2w30C67bf54wpc6qs_jgunQtLXN6XEZ467KaoLtWjlbhIUdc6OeikN-0nfd6lNUgkc6Bk4E' \
  --form profile_image_url=@/home/esteban/Downloads/random_avatar.png
```
expected: 

Expect the body:
```json
{
  "id": 12,
  "firstName": "user1First",
  "last_name": "user1Last",
  "email": "user1First@example.com",
  "email_notification_opt_out": false,
  "profile_image_url": "http://localhost:3000/userprofile-12.png"
}
```

## Resources

- Nestjs modules docs, azure blob nest client

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -41,5 +41,7 @@ lerna-debug.log*
 
 #test data
 **/pgdata**/
+**/public/userprofile**
```
Comment (esteban-gs) [inline .gitignore:44]: This will not commit the images saved here

**Comment by garimacha** [comment]
@esteban-gs  - is this still open for PR? CAn I help review this as it just requires postman calls?

**Comment by esteban-gs** [comment]
Sure! You can test it. It's still in review

File: server/src/file-storage/file-storage.module.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+} from './storage-providers/azure-blob-storage';
+import {
+  LocalStorageProvider,
+  LOCAL_STORAGE_PROVIDER_TOKEN,
+} from './storage-providers/local-storage';
+
+@Module({
+  providers: [
+    { provide: AZURE_BLOB_PROVIDER_TOKEN, useClass: AzureBlobStorage },
+    { provide: LOCAL_STORAGE_PROVIDER_TOKEN, useClass: LocalStorageProvider },
```
Comment (esteban-gs) [inline server/src/file-storage/file-storage.module.ts:15]: This module adds these 2 classes as providers, so the `FileStorageService` can import them. All other modules importing the `FileStorageModule` will get the `FileStorageService` with the providers.

File: server/src/file-storage/file-storage.service.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Arguments to pass to a file storage provider
+ */
+export interface FileStorageArgs {
+  file: Express.Multer.File;
+  userId: number;
+  prefix: FilePrefix;
+  replaceFlag: 'replace' | 'new';
+}
```
Comment (esteban-gs) [inline server/src/file-storage/file-storage.service.ts:14]: This type can be used by any file storage provider

File: server/src/file-storage/file-storage.service.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+   * or other customizations
+   * @param storageArgs
+   * @returns the stored file path
+   */
+  public async storeImage(storageArgs: FileStorageArgs): Promise<string> {
+    if (process.env.MODE === 'production') {
+      return await this.azureStorage.uploadFile(storageArgs);
+    } else {
+      return await this.diskStorage.uploadFile(storageArgs);
+    }
```
Comment (esteban-gs) [inline server/src/file-storage/file-storage.service.ts:47]: Here, we can customize the app to use any implementation of `IFileStorageService`

File: server/src/file-storage/storage-providers/local-storage.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    writableStream.on('error', (err) => {
+      console.log(err.stack);
+    });
+
+    writableStream.write(readableStream);
+
+    return Promise.resolve(devFileUrl);
+  };
+}
```
Comment (esteban-gs) [inline server/src/file-storage/storage-providers/local-storage.ts:28]: This is used in development and it saves files as React assets, which are accessible by `http://localhost:3000/userprofile-{user_id}.{image_extension}`

**Review by jd2rogers2 — APPROVED** [review]
dooooopee

**Pull request merged**